packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -11,10 +11,707 @@ import 'picker.dart';
+class _DatePickerLayout extends MultiChildLayoutDelegate {
+  _DatePickerLayout({
+    @required this.columnWidth,
+    @required this.textDirectionFactor,
+  }) : assert(columnWidth != null),
+       assert(textDirectionFactor != null);
+
+  // The list containing widths of all columns.
+  final List<double> columnWidth;
+
+  // textDirectionFactor is 1 if text is written left to right, and -1 if right to left.
+  final int textDirectionFactor;
+
+  @override
+  void performLayout(Size size) {
+    double remainingWidth = size.width;
+
+    for (int i = 0; i < columnWidth.length; i++)
+      remainingWidth -= columnWidth[i] + _kDatePickerPadSize * 2;
+
+    double currentHorizontalOffset = 0.0;
+
+    for (int i = 0; i < columnWidth.length; i++) {
+      final int index = textDirectionFactor == 1 ? i : columnWidth.length - i - 1;
+
+      double childWidth = columnWidth[index] + _kDatePickerPadSize * 2;
+      if (index == 0 || index == columnWidth.length - 1)
+        childWidth += remainingWidth / 2;
+
+      layoutChild(index, BoxConstraints.tight(new Size(childWidth, size.height)));
+      positionChild(index, new Offset(currentHorizontalOffset, 0.0));
+
+      currentHorizontalOffset += childWidth;
+    }
+  }
+
+  @override
+  bool shouldRelayout(_DatePickerLayout oldDelegate) {
+    return columnWidth != oldDelegate.columnWidth
+      || textDirectionFactor != oldDelegate.textDirectionFactor;
+  }
+}
+
+/// Different modes of [CupertinoDatePicker].
+enum CupertinoDatePickerMode {
+  /// Mode that shows the date in hour, minute, and (optional) an AM/PM designation.
+  ///
+  /// Example: [4 | 14 | PM].
+  time,
+  /// Mode that shows the date in month, day of month, and year.
+  ///
+  /// Example: [July | 13 | 2012].
+  date,
+  /// Mode that shows the date as day of the week, month, day of month and
+  /// the time in hour, minute, and (optional) an AM/PM designation.
+  ///
+  /// Example: [Fri Jul 13 | 4 | 14 | PM]
+  dateAndTime,
+}
+
+/// A date picker widget in iOS style.
+///
+/// There are several modes of the date picker listed in [CupertinoDatePickerMode].
+class CupertinoDatePicker extends StatefulWidget {
+  /// Constructs an iOS style date picker.
+  ///
+  /// [mode] is one of the mode listed in [CupertinoDatePickerMode] and defaults
+  /// to [CupertinoDatePickerMode.dateAndTime].
+  ///
+  /// [onDateTimeChanged] is the callback when the selected date or time changes
+  /// and must not be null.
+  ///
+  /// [initialDateTime] is the initial value of the picker. Must not be null.
+  ///
+  /// [minimumDate] is the minimum date that the picker can be scrolled to in
+  /// [CupertinoDatePickerMode.dateAndTime] mode. Null if there's no limit.
+  ///
+  /// [maximumDate] is the maximum date that the picker can be scrolled to in
+  /// [CupertinoDatePickerMode.dateAndTime] mode. Null if there's no limit.
+  ///
+  /// [minimumYear] is the minimum year that the picker can be scrolled to in
```

Comment:
I want to avoid the situation where user scrolls past a limit and it's not clear what we should do.
For example, if the lower limit is [June | 15 | 2018] and the current picker's value is [February | 19 | 2019]. If user scrolls the year column to 2018, what should we do? There's 2 options:
 - Scroll the year column to 2019.
 - Scroll the picker's value to [June | 15 | 2018].
In my opinion, the second one has usability issue because there's no quick way of undoing the action (user has to reselect all columns).
Therefore, the constraints I set here is only applied to one column.

Reply:
Doesn't this ambiguity apply to month and day too? So we're just partially solving it anyway. 

Since years actually scroll infinitely and month and days only loop, what about just having a min/max year and no min/max date? My general recommendation is to only limit the options that appear (e.g. years) and not force scroll (leave those to the user).